frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts

```diff
@@ -54,3 +56,54 @@ export function calcInitialEditorHeight({
   const lines = getVisibleLinesCount({ query, viewHeight });
   return getEditorLineHeight(lines);
 }
+
+const formatSql = async (sql: string, options: FormatOptionsWithLanguage) => {
+  const sqlFormatter = await import(
+    /* webpackChunkName: "sql-formatter" */ "sql-formatter"
+  );
+  return sqlFormatter.format(sql, options);
+};
+
+const formatterDialectByEngine: Record<string, SqlLanguage> = {
+  "bigquery-cloud-sdk": "bigquery",
+  mysql: "mysql",
+  oracle: "plsql",
+  postgres: "postgresql",
+  "presto-jdbc": "trino",
+  redshift: "redshift",
+  snowflake: "snowflake",
+  sparksql: "spark",
+  sqlite: "sqlite",
+  sqlserver: "tsql",
+};
+
+function getFormatterDialect(engine: string) {
+  if (getEngineNativeType(engine) === "json") {
+    return null;
+  }
+
+  // Fall back to ANSI SQL dialect
+  return formatterDialectByEngine[engine] ?? "sql";
+}
+
+export function canFormatForEngine(engine: string) {
+  return getEngineNativeType(engine) === "sql";
+}
+
+export function formatQuery(queryText: string, engine: string) {
+  const dialect = getFormatterDialect(engine);
+  if (!dialect) {
+    throw new Error(`No formatter dialect for engine ${engine}`);
+  }
+
+  return formatSql(queryText, {
+    language: dialect,
+    tabWidth: 2,
+    keywordCase: "upper",
+    linesBetweenQueries: 2,
+    paramTypes: {
+      // Snippets, parameters, nested questions
```

Comment:
Formatting does not work for queries with [optional clauses](https://www.metabase.com/docs/latest/questions/native-editor/sql-parameters).

Reply:
Good catch! Fixed for engines excluding mssql and sqlite due to the lack of support for it